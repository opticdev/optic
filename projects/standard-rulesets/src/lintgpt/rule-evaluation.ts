import { OpenAI } from 'openai';
import { openai } from './openai';
import { PreparedRule } from './prepare-rule';
import { writeYaml } from '@useoptic/openapi-io';

export type AIRuleEvaluationResult =
  | {
      passed: true;
    }
  | { passed: false; error: string }
  | { skipped: true };
export async function evaluateRule(
  preparedRule: PreparedRule,
  locationContext: string,
  value: any,
  before?: any
): Promise<AIRuleEvaluationResult> {
  const completion = await openai.chat.completions.create({
    messages: [
      {
        role: 'system',
        content: `You are an AI OpenAPI linter that checks if this OpenAPI ${
          preparedRule.entity
        } follows rules the user provides. Location: ${locationContext}.
${
  before
    ? `
Before:
\`\`\`yaml
${writeYaml(value)}
\`\`\`
`
    : ''
}
${before ? 'After' : ''}
\`\`\`yaml
${writeYaml(value)}
\`\`\`

Respond with the returnResult function {passed: true | false, error?: string, skipped: boolean}
If the example is not relevant to the rule ie rule cares about object properties, but provided example is a string property, skipped:true
`,
      },
      {
        role: 'user',
        content: `Does the ${preparedRule.entity} follow this rule?
"${preparedRule.rule}"`,
      },
    ],
    model: 'gpt-4-0613',
    temperature: 0,
    function_call: { name: 'returnResult' },
    functions: [
      {
        name: 'returnResult',
        description: 'The result of the rule',
        parameters: {
          type: 'object',
          required: ['passed', 'skipped'],
          properties: {
            passed: {
              type: 'boolean',
            },
            error: {
              type: 'string',
              description:
                'When passed is false, give the user an error explaining why the rule was failed and why it matters.',
            },
            skipped: {
              type: 'boolean',
              description:
                'The rule is not relevant to the example provided. Skip this result.',
            },
          },
        },
      },
    ],
  });

  // console.log(JSON.stringify(completion, null, 2));

  try {
    const functionCall = completion.choices[0].message.function_call!;

    const parsed = JSON.parse(functionCall.arguments);
    if (parsed.skipped) {
      return { skipped: true };
    }
    if (parsed.passed) {
      return {
        passed: true,
      };
    } else {
      return {
        passed: parsed.passed,
        error: parsed.error,
      };
    }
  } catch (e) {
    throw new Error('Error evaluating lintgpt rule ' + e);
  }
}
