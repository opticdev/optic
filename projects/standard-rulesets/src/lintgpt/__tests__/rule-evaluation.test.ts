import { describe, expect, jest, test } from '@jest/globals';
import { evaluateRule } from '../rule-evaluation';
import { OpenAPIV3 } from 'openapi-types';

jest.setTimeout(10000);

describe('evaluate rules:', () => {
  test('Check for operationId -- should pass', async () => {
    const operation: OpenAPIV3.OperationObject = {
      operationId: 'getUsers',
      responses: {
        '200': {
          description: 'all the users',
        },
      },
    };
    const result = await evaluateRule(
      {
        entity: 'OPERATION',
        rule: 'Each operation must have an operationId',
        changed: false,
        severity: 'ERROR',
      },
      'GET /todos',
      operation
    );

    expect(result).toMatchSnapshot();
  });
  test('Check for operationId -- should fail', async () => {
    const operation: OpenAPIV3.OperationObject = {
      responses: {
        '200': {
          description: 'all the users',
        },
      },
    };
    const result = await evaluateRule(
      {
        entity: 'OPERATION',
        rule: 'Each operation must have an operationId',
        changed: false,
        severity: 'ERROR',
      },
      'GET /todos',
      operation
    );

    expect(result).toMatchSnapshot();
  });

  describe('Fuzzy property naming rules', () => {
    test('property with date-like name -- should fail', async () => {
      const property = {
        created_at: {
          type: 'string',
          description: 'when the order was made',
        },
      };
      const result = await evaluateRule(
        {
          entity: 'PROPERTY',
          rule: 'Properties that sound like they contain a dates should use format: datetime',
          changed: false,
          severity: 'WARNING',
        },
        'GET /todos in the 200 response',
        property
      );
      expect(result).toMatchSnapshot();
    });
    test('property with date-like name -- should pass', async () => {
      const property = {
        created_at: {
          type: 'string',
          format: 'datetime',
          description: 'when the order was made',
        },
      };
      const result = await evaluateRule(
        {
          entity: 'PROPERTY',
          rule: 'Properties that sound like they contain a dates should use format: datetime',
          changed: false,
          severity: 'WARNING',
        },
        'GET /todos in the 200 response',
        property
      );
      expect(result).toMatchSnapshot();
    });
    test('The keys in an object property should all follow the same naming convention -- not relevant', async () => {
      const property = {
        userId: {
          type: 'string',
        },
      };
      const result = await evaluateRule(
        {
          entity: 'PROPERTY',
          rule: 'The keys in an object property should all follow the same naming convention',
          changed: false,
          severity: 'WARNING',
        },
        'GET /todos in the 200 response',
        property
      );
      expect(result).toMatchSnapshot();
    });
    test('The keys in an object property should all follow the same naming convention -- should fail', async () => {
      const property = {
        user: {
          ID: {
            type: 'string',
          },
          fullName: {
            type: 'string',
          },
          last_name: {
            type: 'string',
          },
        },
      };
      const result = await evaluateRule(
        {
          entity: 'PROPERTY',
          rule: 'The keys in an object property should all follow the same naming convention',
          changed: false,
          severity: 'WARNING',
        },
        'GET /todos in the 200 response',
        property
      );
      expect(result).toMatchSnapshot();
    });
  });
});
