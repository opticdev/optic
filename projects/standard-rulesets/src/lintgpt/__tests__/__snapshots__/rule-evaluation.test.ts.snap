// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluate rules: Check for operationId -- should fail 1`] = `
{
  "error": "The operation does not have an operationId.",
  "passed": false,
}
`;

exports[`evaluate rules: Check for operationId -- should pass 1`] = `
{
  "passed": true,
}
`;

exports[`evaluate rules: Fuzzy property naming rules The keys in an object property should all follow the same naming convention -- not relevant 1`] = `
{
  "skipped": true,
}
`;

exports[`evaluate rules: Fuzzy property naming rules The keys in an object property should all follow the same naming convention -- should fail 1`] = `
{
  "error": "The keys in the object property do not follow the same naming convention. 'ID' and 'last_name' are not consistent.",
  "passed": false,
}
`;

exports[`evaluate rules: Fuzzy property naming rules property with date-like name -- should fail 1`] = `
{
  "error": "The 'created_at' property sounds like it contains a date but it does not use the 'format: datetime'.",
  "passed": false,
}
`;

exports[`evaluate rules: Fuzzy property naming rules property with date-like name -- should pass 1`] = `
{
  "passed": true,
}
`;
